File: dp07_countContruct.go
package main

import "strings"

func main() {
	// println(countConstruct("abcdef", []string{"ab", "abc", "cd", "def", "abcd"}))
	// println(countConstruct("skateboard", []string{"bo", "rd", "ate", "t", "ska", "sk", "boar"}))
	// println(countConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeef", []string{"e", "ee", "eee", "eeee", "eeeee", "eeeeeee", "eeeeeeee"}))
	println(countConstructDy("abcdef", []string{"ab", "abc", "cd", "def", "abcd"}, map[string]int{}))
	println(countConstructDy("skateboard", []string{"bo", "rd", "ate", "t", "ska", "sk", "boar"}, map[string]int{}))
	println(countConstructDy("eeeeeeeeeeeeeeeeeeeeeeeeeeeef", []string{"e", "ee", "eee", "eeee", "eeeee", "eeeeeee", "eeeeeeee"}, map[string]int{}))
}

func countConstructDy(target string, wordBank []string, memo map[string]int) int {
	if val, ok := memo[target]; ok {
		return val
	}

	if target == "" {
		return 1
	}

	var count int

	for _, word := range wordBank {
		if strings.Index(target, word) == 0 {
			suffix := target[len(word):]
			noOfWays := countConstructDy(suffix, wordBank, memo)
			count += noOfWays
		}
	}

	memo[target] = count
	return count
}

func countConstruct(target string, wordBank []string) int {
	if target == "" {
		return 1
	}

	var count int

	for _, word := range wordBank {
		if strings.Index(target, word) == 0 {
			suffix := target[len(word):]
			noOfWays := countConstruct(suffix, wordBank)
			count += noOfWays
		}
	}

	return count
}

File: dp07_count_contruct.py
def count_construct(target, word_banks):
    if target == "":
        return 1

    count = 0

    for word in word_banks:
        if target.find(word) == 0:
            suffix = target[len(word):]
            no_of_ways = count_construct(suffix, word_banks)
            count += no_of_ways

    return count


def count_construct_dy(target, word_banks, memo):
    if target in memo:
        return memo[target]

    if target == "":
        return 1

    count = 0

    for word in word_banks:
        if target.find(word) == 0:
            suffix = target[len(word):]
            no_of_ways = count_construct_dy(suffix, word_banks, memo)
            count += no_of_ways

    memo[target] = count
    return count

def main():
    # print(count_construct('abcdef', ['ab', 'abc', 'cd', 'def', 'abcd']))
    # print(count_construct('skateboard',  ['bo', 'rd', 'ate', 't', 'ska', 'sk', 'boar']))
    # print(count_construct('eeeeeeeeeeeeeeeeeeeeeeeeef',  ['e', 'ee', 'eee', 'eeee', 'eeeee', 'eeeeeee', 'eeeeeeee']))
    
    print(count_construct_dy('abcdef', ['ab', 'abc', 'cd', 'def', 'abcd'], {}))
    print(count_construct_dy('skateboard',  ['bo', 'rd', 'ate', 't', 'ska', 'sk', 'boar'], {}))
    print(count_construct_dy('eeeeeeeeeeeeeeeeeeeeeeeeeee',  ['e', 'ee', 'eee', 'eeee', 'eeeee', 'eeeeeee', 'eeeeeeee'], {}))

main()